src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -0,0 +1,391 @@
+        shell=shell,
+        xdg_token=xdg_token,
+        client=client,
+        focus_controller=focus_controller]
+    {
+        auto const wl_surface = mf::WlSurface::from(surface);
+        if (!wl_surface)
+        {
+            mir::log_error("XdgActivationV1::activate wl_surface not found");
+            return;
+        }
+
+        auto const scene_surface_opt = wl_surface->scene_surface();
+        if (!scene_surface_opt)
+        {
+            mir::log_error("XdgActivationV1::activate scene_surface_opt not found");
+            return;
+        }
+
+        // In the event of a failure, we send 'done' with a new, invalid token which cannot be used later.
+        if (xdg_token->seat && xdg_token->serial)
+        {
+            // First, assert that the seat still exists.
+            auto const wl_seat = mf::WlSeat::from(xdg_token->seat.value());
+            if (!wl_seat)
+                mir::log_warning("XdgActivationTokenV1::activate wl_seat not found");
+
+            auto event = client->event_for(xdg_token->serial.value());
+            if (event == std::nullopt)
+                mir::log_warning("XdgActivationTokenV1::activate serial not found");
+        }
+
+        if (xdg_token->app_id)
+        {
+            // Check that the focused application has app_id
+            auto const& focused = focus_controller->focused_surface();
+            if (!focused)
+            {
+                mir::log_error("XdgActivationTokenV1::activate cannot find the focused surface");
+                return;
+            }
+
+            if (focused->application_id() != xdg_token->app_id)
+            {
+                mir::log_error("XdgActivationTokenV1::activate app_id != focused application id");
+                return;
+            }
+        }
+
+        if (xdg_token->requesting_surface)
+        {
+            // Check that the focused surface is the requesting_surface
+            auto const requesting_wl_surface = mf::WlSurface::from(xdg_token->requesting_surface.value());
+            auto const requesting_scene_surface_opt = requesting_wl_surface->scene_surface();
+            if (!requesting_scene_surface_opt)
+            {
+                mir::log_error("XdgActivationTokenV1::commit cannot find the provided scene surface");
+                return;
+            }
+
+            auto const& scene_surface = requesting_scene_surface_opt.value();
+            if (focus_controller->focused_surface() != scene_surface)
+            {
+                mir::log_error("XdgActivationTokenV1::commit the focused surface is not the surface that requested activation");
+                return;
+            }
+        }
+
+        auto const& scene_surface = scene_surface_opt.value();
+        auto const now = std::chrono::steady_clock::now().time_since_epoch();
+        auto ns = std::chrono::duration_cast<std::chrono::nanoseconds >(now).count();
+        shell->raise_surface(wl_surface->session, scene_surface, ns);
```

Comment:
I think this should be a request _via_ the window manager, not bypassing the window manager.

In particular, the WM may want to switch workspaces or have a reason (screen lock?) to ignore the request.

Reply:
As we've discussed, what I've done is the right thing :+1: